casper/src/main/scala/coop/rchain/casper/util/comm/CommUtil.scala

```diff
@@ -52,41 +56,117 @@ object CommUtil {
+    : F[Unit] = {
+    val request = ApprovedBlockRequest("PleaseSendMeAnApprovedBlock").toByteString
+
+    def askPeers(peers: List[PeerNode], local: PeerNode): F[Unit] = peers match {
+      case peer :: rest =>
+        for {
+          _    <- Log[F].info(s"Sending request for ApprovedBlock to $peer")
+          send <- TransportLayer[F].roundTrip(peer, packet(local, request), 5.seconds)
+          _ <- send match {
+                case Left(err) =>
+                  Log[F].info(s"Failed to get response from $peer because: $err") *>
+                    askPeers(rest, local)
+
+                case Right(response) =>
+                  Log[F].info(s"Received response from $peer! Dispatching...")
+                  Connect
+                    .dispatch[F](response)
+                    .flatMap(_ => {
+                      MultiParentCasperConstructor[F].lastApprovedBlock.map {
+                        case Some(_) => () //got one!
+                        case None    => askPeers(rest, local)
+                      }
+                    })
+
+              }
+        } yield ()
+
+      case Nil => ().pure[F]
+    }
+
+    for {
+      a     <- MultiParentCasperConstructor[F].lastApprovedBlock
+      peers <- NodeDiscovery[F].peers
+      local <- TransportLayer[F].local
+      _ <- if (a.isEmpty) askPeers(peers.toList, local)
+          else ().pure[F]
+    } yield ()
+  }
+
   def casperPacketHandler[
-      F[_]: Monad: MultiParentCasper: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler](
+      F[_]: Monad: MultiParentCasperConstructor: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler](
       peer: PeerNode): PartialFunction[Packet, F[Option[Packet]]] =
     Function
       .unlift(
-        (p: Packet) => { packetToBlockRequest(p) orElse packetToBlockMessage(p) }
+        (p: Packet) => {
+          packetToBlockRequest(p) orElse packetToApprovedBlock(p) orElse packetToApprovedBlockRequest(
+            p) orElse packetToBlockMessage(p)
+        }
       )
       .andThen {
-        case b: BlockMessage =>
-          for {
-            isOldBlock <- MultiParentCasper[F].contains(b)
-            _ <- if (isOldBlock) {
-                  Log[F].info(
-                    s"CASPER: Received block ${PrettyPrinter.buildString(b.blockHash)} again.")
-                } else {
-                  handleNewBlock[F](b)
-                }
-          } yield none[Packet]
-
-        case r: BlockRequest =>
+        case b @ (_: BlockMessage | _: BlockRequest) =>
+          MultiParentCasperConstructor[F].casperInstance match {
+            case Left(ex) =>
+              Log[F]
+                .warn(
+                  "CASPER: a Casper message was received, " +
+                    "however could not be processed due to the following error. " +
+                    ex.getMessage
+                )
+                .map(_ => none[Packet])
```

Comment:
Is this equivalent to

`Log[F].warn(...) *> Applicative[F].pure(none[Packet])`?

Reply:
Yes. Though, in general `*>` is different from `map` because the former sequences unrelated effects, where as the latter transforms values within the same effect. From a style perspective I'm not sure which is better, to have an sequence into a `pure` or an unconditional function into the desired value.

Reply:
`*>` is eager where `map/flatMap` is lazy with trampolined monads.